File: validation/open-immersive/aurora_oar_5_1_differential_probe.c
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "oar.h"
#include "test_helpers.h"

#define TEST_SAMPLE_RATE 48000u
#define TEST_FRAMES 256u
#define TEST_FREQUENCY_HZ 440.0f
#define TEST_ELEMENT_ID 1u
#define TEST_CHANNELS 6u

typedef struct {
  const char *name;
  const char *semantic_position;
  float azimuth_degrees;
  float gain_db;
} case_spec_t;

typedef struct {
  double channel_levels[TEST_CHANNELS];
  int finite;
} case_result_t;

static oar_metadata_t *create_gain_metadata(uint32_t gain_id,
                                            float gain_value_db,
                                            uint32_t samples_per_channel) {
  oar_metadata_t *metadata = (oar_metadata_t *)malloc(sizeof(oar_metadata_t));
  if (!metadata) return NULL;
  memset(metadata, 0, sizeof(*metadata));
  metadata->type = ck_metadata_gain;
  metadata->duration = (int)samples_per_channel;
  metadata->gain.id = gain_id;
  metadata->gain.param_type = ck_param_constant;
  metadata->gain.constant_gain = gain_value_db;
  return metadata;
}

static double channel_rms(const float *samples, uint32_t count, int *finite) {
  double sum_squares = 0.0;
  for (uint32_t i = 0; i < count; ++i) {
    double sample = (double)samples[i];
    if (!isfinite(sample)) *finite = 0;
    sum_squares += sample * sample;
  }
  return sqrt(sum_squares / (double)count);
}

static int render_case(const case_spec_t *spec, case_result_t *result) {
  oar_config_t config =
      create_config(ck_oar_layout_51, TEST_FRAMES, TEST_SAMPLE_RATE);
  oar_t *oar = oar_create(&config);
  if (!oar) {
    fprintf(stderr, "oar_create failed for %s\n", spec->name);
    return -1;
  }

  int rc = -1;
  int group_id = oar_add_audio_group(oar);
  if (group_id < 0) {
    fprintf(stderr, "oar_add_audio_group failed for %s\n", spec->name);
    goto cleanup_oar;
  }

  oar_audio_element_config_t element_config = create_object_element_config(1);
  if (oar_add_audio_element(oar, group_id, TEST_ELEMENT_ID, &element_config) !=
      0) {
    fprintf(stderr, "oar_add_audio_element failed for %s\n", spec->name);
    goto cleanup_oar;
  }

  uint32_t samples_per_channel = oar_get_samples_per_channel(oar);
  uint32_t output_channels = oar_get_number_of_output_channels(oar);
  if (samples_per_channel != TEST_FRAMES || output_channels != TEST_CHANNELS) {
    fprintf(stderr,
            "unexpected OAR 5.1 dimensions for %s: frames=%u channels=%u\n",
            spec->name, samples_per_channel, output_channels);
    goto cleanup_oar;
  }

  oar_audio_block_t input;
  if (alloc_audio_block(1, samples_per_channel, &input) != 0) {
    fprintf(stderr, "input allocation failed for %s\n", spec->name);
    goto cleanup_oar;
  }
  generate_sine(input.data, samples_per_channel, TEST_FREQUENCY_HZ,
                (float)TEST_SAMPLE_RATE);
  if (oar_update_audio_element_data(oar, TEST_ELEMENT_ID, &input) != 0) {
    fprintf(stderr, "oar_update_audio_element_data failed for %s\n",
            spec->name);
    free(input.data);
    goto cleanup_oar;
  }
  free(input.data);

  polar_t position = {spec->azimuth_degrees, 0.0f, 1.0f};
  if (set_object_position(oar, TEST_ELEMENT_ID, &position,
                          samples_per_channel) != 0) {
    fprintf(stderr, "position metadata update failed for %s\n", spec->name);
    goto cleanup_oar;
  }

  if (spec->gain_db != 0.0f) {
    oar_metadata_t *gain =
        create_gain_metadata(1, spec->gain_db, samples_per_channel);
    if (!gain) {
      fprintf(stderr, "gain metadata allocation failed for %s\n", spec->name);
      goto cleanup_oar;
    }
    int gain_rc =
        oar_update_audio_element_metadata(oar, TEST_ELEMENT_ID, gain);
    free(gain);
    if (gain_rc != 0) {
      fprintf(stderr, "gain metadata update failed for %s\n", spec->name);
      goto cleanup_oar;
    }
  }

  oar_audio_block_t output;
  if (alloc_audio_block(output_channels, samples_per_channel, &output) != 0) {
    fprintf(stderr, "output allocation failed for %s\n", spec->name);
    goto cleanup_oar;
  }
  if (oar_render(oar, &output) != 0) {
    fprintf(stderr, "oar_render failed for %s\n", spec->name);
    free(output.data);
    goto cleanup_oar;
  }

  result->finite = 1;
  for (uint32_t channel = 0; channel < TEST_CHANNELS; ++channel) {
    result->channel_levels[channel] = channel_rms(
        output.data + channel * samples_per_channel, samples_per_channel,
        &result->finite);
    if (!isfinite(result->channel_levels[channel])) result->finite = 0;
  }
  free(output.data);
  rc = 0;

cleanup_oar:
  oar_destroy(oar);
  return rc;
}

int main(int argc, char **argv) {
  if (argc > 2) {
    fprintf(stderr, "usage: %s [OUTPUT_JSON]\n", argv[0]);
    return 64;
  }

  const case_spec_t specs[] = {
      {"front_left_unity", "front_left", 30.0f, 0.0f},
      {"center_unity", "center", 0.0f, 0.0f},
      {"front_right_unity", "front_right", -30.0f, 0.0f},
      {"surround_left_unity", "surround_left", 110.0f, 0.0f},
      {"surround_right_unity", "surround_right", -110.0f, 0.0f},
      {"center_minus_6db", "center", 0.0f, -6.0f},
  };
  const size_t case_count = sizeof(specs) / sizeof(specs[0]);
  case_result_t results[sizeof(specs) / sizeof(specs[0])];

  for (size_t i = 0; i < case_count; ++i) {
    if (render_case(&specs[i], &results[i]) != 0) return 1;
  }

  FILE *out = stdout;
  if (argc == 2) {
    out = fopen(argv[1], "w");
    if (!out) {
      perror("fopen");
      return 1;
    }
  }

  fprintf(out,
          "{\n"
          "  \"schema_version\": 1,\n"
          "  \"implementation\": \"aomedia-oar-1.0.0\",\n"
          "  \"layout\": \"5.1\",\n"
          "  \"sample_rate\": %u,\n"
          "  \"frames_per_case\": %u,\n"
          "  \"channel_order\": [\"FL\", \"FR\", \"FC\", \"LFE\", \"SL\", \"SR\"],\n"
          "  \"coordinate_convention\": \"positive-azimuth-left\",\n"
          "  \"lfe_semantics\": \"non-directional-zero-for-object-render\",\n"
          "  \"cases\": [\n",
          TEST_SAMPLE_RATE, TEST_FRAMES);

  for (size_t i = 0; i < case_count; ++i) {
    fprintf(out,
            "    {\"name\": \"%s\", \"semantic_position\": \"%s\", "
            "\"azimuth_degrees\": %.6f, \"gain_db\": %.6f, "
            "\"frame_count\": %u, \"channel_levels\": [%.12g, %.12g, %.12g, %.12g, %.12g, %.12g], "
            "\"finite\": %s}%s\n",
            specs[i].name, specs[i].semantic_position, specs[i].azimuth_degrees,
            specs[i].gain_db, TEST_FRAMES, results[i].channel_levels[0],
            results[i].channel_levels[1], results[i].channel_levels[2],
            results[i].channel_levels[3], results[i].channel_levels[4],
            results[i].channel_levels[5], results[i].finite ? "true" : "false",
            i + 1 == case_count ? "" : ",");
  }
  fprintf(out, "  ]\n}\n");

  if (out != stdout && fclose(out) != 0) {
    perror("fclose");
    return 1;
  }
  fprintf(stderr, "AURORA-OAR-5-1-DIFFERENTIAL-PROBE-PASS\n");
  return 0;
}

File: crates/aurora-simulation-assurance/src/bin/aurora-oar-5-1-differential-probe.rs
use std::f32::consts::PI;
use std::fs;
use std::path::PathBuf;

use anyhow::{bail, Context, Result};
use aurora_core::{ChannelRole, Listener, Speaker, Vector3};
use aurora_renderer_api::{RenderObject, Renderer, RendererScratch, SpeakerGain};
use aurora_renderer_vbap::ObjectVbapRenderer;
use clap::Parser;
use serde::Serialize;

const SAMPLE_RATE: u32 = 48_000;
const FRAMES: usize = 256;
const FREQUENCY_HZ: f32 = 440.0;
const CHANNELS: usize = 6;

#[derive(Debug, Parser)]
#[command(name = "aurora-oar-5-1-differential-probe")]
#[command(about = "Emit deterministic Aurora 5.1 object semantics for pinned OAR comparison")]
struct Args {
    #[arg(long)]
    output: PathBuf,
}

#[derive(Debug, Serialize)]
struct ProbeReport {
    schema_version: u32,
    implementation: &'static str,
    layout: &'static str,
    sample_rate: u32,
    frames_per_case: usize,
    channel_order: [&'static str; CHANNELS],
    coordinate_convention: &'static str,
    lfe_semantics: &'static str,
    cases: Vec<ProbeCase>,
}

#[derive(Debug, Serialize)]
struct ProbeCase {
    name: &'static str,
    semantic_position: &'static str,
    azimuth_degrees: f32,
    gain_db: f32,
    frame_count: usize,
    channel_levels: [f64; CHANNELS],
    finite: bool,
}

#[derive(Debug, Clone, Copy)]
struct CaseSpec {
    name: &'static str,
    semantic_position: &'static str,
    azimuth_degrees: f32,
    gain_db: f32,
}

fn main() -> Result<()> {
    let args = Args::parse();
    let report = run_probe()?;
    if let Some(parent) = args.output.parent() {
        fs::create_dir_all(parent)
            .with_context(|| format!("failed to create {}", parent.display()))?;
    }
    fs::write(
        &args.output,
        serde_json::to_vec_pretty(&report).context("failed to encode 5.1 probe report")?,
    )
    .with_context(|| format!("failed to write {}", args.output.display()))?;
    println!("AURORA-OAR-5-1-DIFFERENTIAL-PROBE-PASS");
    println!("report={}", args.output.display());
    Ok(())
}

fn run_probe() -> Result<ProbeReport> {
    let specs = [
        CaseSpec {
            name: "front_left_unity",
            semantic_position: "front_left",
            azimuth_degrees: -30.0,
            gain_db: 0.0,
        },
        CaseSpec {
            name: "center_unity",
            semantic_position: "center",
            azimuth_degrees: 0.0,
            gain_db: 0.0,
        },
        CaseSpec {
            name: "front_right_unity",
            semantic_position: "front_right",
            azimuth_degrees: 30.0,
            gain_db: 0.0,
        },
        CaseSpec {
            name: "surround_left_unity",
            semantic_position: "surround_left",
            azimuth_degrees: -110.0,
            gain_db: 0.0,
        },
        CaseSpec {
            name: "surround_right_unity",
            semantic_position: "surround_right",
            azimuth_degrees: 110.0,
            gain_db: 0.0,
        },
        CaseSpec {
            name: "center_minus_6db",
            semantic_position: "center",
            azimuth_degrees: 0.0,
            gain_db: -6.0,
        },
    ];

    let mut cases = Vec::with_capacity(specs.len());
    for spec in specs {
        cases.push(render_case(spec)?);
    }

    Ok(ProbeReport {
        schema_version: 1,
        implementation: "aurora-object-vbap2d",
        layout: "5.1",
        sample_rate: SAMPLE_RATE,
        frames_per_case: FRAMES,
        channel_order: ["FL", "FR", "FC", "LFE", "SL", "SR"],
        coordinate_convention: "negative-azimuth-left",
        lfe_semantics: "non-directional-zero-for-object-render",
        cases,
    })
}

fn render_case(spec: CaseSpec) -> Result<ProbeCase> {
    let mut renderer = ObjectVbapRenderer::new();
    renderer
        .configure(five_one_speakers(), SAMPLE_RATE, FRAMES, 1)
        .context("failed to configure Aurora object-safe 5.1 renderer")?;
    if renderer.output_channel_count() != CHANNELS {
        bail!(
            "Aurora 5.1 renderer returned {} channels",
            renderer.output_channel_count()
        );
    }

    let listener = Listener {
        position: Vector3::ZERO,
        orientation: Vector3::new(1.0, 0.0, 0.0),
        ear_height: 0.0,
    };
    let object = RenderObject {
        position: vector_from_azimuth(spec.azimuth_degrees),
        gain: db_to_linear(spec.gain_db),
    };
    let mut gains = [SpeakerGain::default(); CHANNELS];
    let mut scratch = RendererScratch::new(
        renderer
            .required_scratch_size()
            .context("failed to obtain Aurora 5.1 scratch size")?,
    );
    renderer
        .render_gains(&listener, &[object], &mut gains, &mut scratch)
        .context("Aurora 5.1 object render failed")?;

    for (index, gain) in gains.iter().enumerate() {
        if gain.speaker_index != index {
            bail!(
                "unexpected Aurora 5.1 channel order at {index}: {}",
                gain.speaker_index
            );
        }
    }

    let finite = gains.iter().all(|gain| {
        gain.gain.is_finite() && gain.distance_meters.is_finite() && gain.delay_samples.is_finite()
    });
    let channel_levels = std::array::from_fn(|index| rms_for_gain(gains[index].gain));
    let finite = finite && channel_levels.iter().all(|level| level.is_finite());

    Ok(ProbeCase {
        name: spec.name,
        semantic_position: spec.semantic_position,
        azimuth_degrees: spec.azimuth_degrees,
        gain_db: spec.gain_db,
        frame_count: FRAMES,
        channel_levels,
        finite,
    })
}

fn five_one_speakers() -> Vec<Speaker> {
    vec![
        speaker("FL", "Front Left", ChannelRole::FrontLeft, -30.0),
        speaker("FR", "Front Right", ChannelRole::FrontRight, 30.0),
        speaker("FC", "Front Center", ChannelRole::FrontCenter, 0.0),
        speaker(
            "LFE",
            "Low Frequency Effects",
            ChannelRole::LowFrequencyEffects,
            0.0,
        ),
        speaker("SL", "Surround Left", ChannelRole::SurroundLeft, -110.0),
        speaker("SR", "Surround Right", ChannelRole::SurroundRight, 110.0),
    ]
}

fn speaker(id: &str, label: &str, role: ChannelRole, azimuth_degrees: f32) -> Speaker {
    Speaker {
        id: id.to_owned(),
        label: label.to_owned(),
        channel_role: role,
        position: vector_from_azimuth(azimuth_degrees),
        orientation: Vector3::ZERO,
        gain_db: 0.0,
        delay_samples: 0.0,
        enabled: true,
    }
}

fn vector_from_azimuth(degrees: f32) -> Vector3 {
    let radians = degrees * PI / 180.0;
    Vector3::new(radians.cos(), radians.sin(), 0.0)
}

fn db_to_linear(db: f32) -> f32 {
    10.0_f32.powf(db / 20.0)
}

fn rms_for_gain(gain: f32) -> f64 {
    let mut sum_squares = 0.0_f64;
    for frame in 0..FRAMES {
        let phase = 2.0 * PI * FREQUENCY_HZ * frame as f32 / SAMPLE_RATE as f32;
        let sample = phase.sin() * gain;
        sum_squares += f64::from(sample) * f64::from(sample);
    }
    (sum_squares / FRAMES as f64).sqrt()
}
